### PR TITLE
initial support for --close-pr (REVIEW)

### PR DESCRIPTION
**Pull request by migueldiascosta**

### CODE REVIEW

File: easybuild/tools/github.py [hunk 1]
```diff
@@ -958,6 +958,53 @@ def not_eligible(msg):
...
+        raise EasyBuildError("PR #%d from %s/%s is already closed.", pr, pr_target_account, pr_target_repo)
+
+    msg = "\n%s/%s PR #%s was submitted by %s, " % (pr_target_account, pr_target_repo, pr, pr_data['user']['login'])
+    msg += "you are using GitHub account '%s'\n" % github_user
+    print_msg(msg, prefix=False)
+
+    dry_run = build_option('dry_run') or build_option('extended_dry_run')
+
+    if not reason:
+        reason = "unspecified"
```
Comment (boegel) [inline easybuild/tools/github.py]: Maybe make this `"(no reason specified)"`?

File: easybuild/tools/github.py [hunk 2]
```diff
@@ -958,6 +958,53 @@ def not_eligible(msg):
...
+
+    comment = "@%s, this PR is being closed for the following reason: %s." % (pr_data['user']['login'], reason)
+    comment += " Do reopen if the PR is still relevant."
```
Comment (boegel) [inline easybuild/tools/github.py]: `Please don't hesitate to reopen this PR or add a comment if you feel this contribution is still relevant.`

Comment (migueldiascosta) [reply easybuild/tools/github.py]: and a link to the new section in the documentation?

Comment (boegel) [reply easybuild/tools/github.py]: Or better `our policy`

**Comment by migueldiascosta** [comment]
there are quite a few ways this could be improved (e.g., in dry-run mode, show related PRs, whether it uses archived toolchain, date of last commit and comments, etc.), but that would involve even more code duplication with `merge_pr` and `review_pr`...

not sure if the best approach would be a common close/merge function or fleshing out (almost) identical code blocks...

File: easybuild/tools/github.py [hunk 3]
```diff
@@ -958,6 +958,55 @@ def not_eligible(msg):
...
+    pr_url = lambda g: g.repos[pr_target_account][pr_target_repo].pulls[pr]
+    status, pr_data = github_api_get_request(pr_url, github_user)
+    if status != HTTP_STATUS_OK:
+        raise EasyBuildError("Failed to get data for PR #%d from %s/%s (status: %d %s)",
+                             pr, pr_target_account, pr_target_repo, status, pr_data)
+
+    if pr_data['state'] == 'closed':
+        raise EasyBuildError("PR #%d from %s/%s is already closed.", pr, pr_target_account, pr_target_repo)
+
+    msg = "\n%s/%s PR #%s was submitted by %s, " % (pr_target_account, pr_target_repo, pr, pr_data['user']['login'])
```
Comment (boegel) [inline easybuild/tools/github.py]: introduce an additional local variable `pr_owner` (or similar) to simplify things a bit:

```python
pr_owner = pr_data['user']['login']
```

File: easybuild/tools/github.py [hunk 4]
```diff
@@ -958,6 +958,55 @@ def not_eligible(msg):
...
+        raise EasyBuildError("GitHub user must be specified to use --close-pr")
+
+    pr_target_account = build_option('pr_target_account')
+    pr_target_repo = build_option('pr_target_repo')
+
+    pr_url = lambda g: g.repos[pr_target_account][pr_target_repo].pulls[pr]
+    status, pr_data = github_api_get_request(pr_url, github_user)
+    if status != HTTP_STATUS_OK:
+        raise EasyBuildError("Failed to get data for PR #%d from %s/%s (status: %d %s)",
+                             pr, pr_target_account, pr_target_repo, status, pr_data)
```
Comment (boegel) [inline easybuild/tools/github.py]: grabbing `pr_data` is becoming pretty common, we should introduce a `fetch_pr_data` function that we can call where relevant, code duplication is indeed becoming a bit much in `github.py`...

File: easybuild/tools/github.py [hunk 5]
```diff
@@ -958,6 +958,55 @@ def not_eligible(msg):
...
+    if dry_run:
+        print_msg("[DRY RUN] Closed %s/%s pull request #%s" % (pr_target_account, pr_target_repo, pr), prefix=False)
+    else:
+        github_token = fetch_github_token(github_user)
+        if github_token is None:
+            raise EasyBuildError("GitHub token for user '%s' must be available to use --close-pr", github_user)        
+        g = RestClient(GITHUB_API_URL, username=github_user, token=github_token)
+        pull_url = g.repos[pr_target_account][pr_target_repo].pulls[pr]
+        body = {'state': 'closed'}
+        status, data = pull_url.post(body=body)
```
Comment (boegel) [inline easybuild/tools/github.py]: There's a lot of duplication here too, maybe time for a `pr_action` function that gets PR# + `body` as arguments (and `github_token` as optional argument)?

File: easybuild/tools/github.py [hunk 6]
```diff
@@ -1487,3 +1521,12 @@ def find_easybuild_easyconfig(github_user=None):
 
     eb_file = os.path.join(eb_parent_path, fn)
     return eb_file
+
```
Comment (boegel) [inline easybuild/tools/github.py]: @migueldiascosta Style nitpicking: top-level functions are separated with double empty line

File: easybuild/tools/github.py [hunk 7]
```diff
@@ -1487,3 +1521,12 @@ def find_easybuild_easyconfig(github_user=None):
...
+def fetch_pr_data(pr, pr_target_account, pr_target_repo, github_user):
```
Comment (boegel) [inline easybuild/tools/github.py]: @migueldiascosta Should we add a dedicated test for `fetch_pr_data` in `test/framework/github.py`?

File: easybuild/tools/github.py (hunk 5 above)
Comment (boegel) [reply easybuild/tools/github.py]: @migueldiascosta Does this suggestion still make sense, or not?

Comment (migueldiascosta) [reply easybuild/tools/github.py]: @boegel I don't know - each case is slightly different, but I may be missing an obvious generalization

File: easybuild/tools/github.py (hunk 7 above)
Comment (migueldiascosta) [reply easybuild/tools/github.py]: @boegel will do

**Comment by boegel** [comment]
In the context of considering auto-closing of stale PRs (e.g. through @boegelbot), there are quite a bit of good/interesting ideas/viewpoints in https://github.com/spack/spack/pull/8811 .

The main difference I guess is that our approach would be project-aware (or at least it could be).

**Comment by boegel** [comment]
@migueldiascosta Needs a sync with `develop` to fix merge conflict.

Other than that, do you think this is ready to go?

**Comment by boegel** [comment]
@migueldiascosta Let's sync this branch with current `develop` now that #2400 is merged?

**Comment by boegel** [comment]
@migueldiascosta More conflicts, probably because of #2668?

File: easybuild/tools/github.py (hunk 5 above)
Comment (boegel) [reply easybuild/tools/github.py]: Just took a look myself, it's indeed not entirely trivial, so let's leave it as is (for now).

**Comment by boegel** [comment]
@migueldiascosta Some code cleanup in https://github.com/migueldiascosta/easybuild-framework/pull/6 (tests still pass).

One thing that is still missing is a dedicated test for `close_pr` and `reasons_for_closing`, but I'm not sure that's worth the trouble (certainly not for `reasons_for_closing`).

**Comment by boegel** [comment]
@migueldiascosta Tests added in https://github.com/migueldiascosta/easybuild-framework/pull/7

**Comment by boegel** [comment]
lgtm, `github` tests pass, so going in, thanks a lot @migueldiascosta!

**Pull request merged**